src/main/java/org/opentripplanner/routing/graph_finder/PlaceFinderTraverseVisitor.java

```diff
@@ -0,0 +1,175 @@
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * A TraverseVisitor used in finding various types of places while walking the street graph.
+ *
+ * TODO Add car and bike parks
+ */
+public class PlaceFinderTraverseVisitor implements TraverseVisitor {
+
+  public final List<PlaceAtDistance> placesFound = new ArrayList<>();
+  private final RoutingService routingService;
+  private final Set<TransitMode> filterByModes;
+  private final Set<FeedScopedId> filterByStops;
+  private final Set<FeedScopedId> filterByRoutes;
+  private final Set<String> filterByBikeRentalStation;
+  private final Set<String> seenPatternAtStops = new HashSet<>();
+  private final Set<FeedScopedId> seenStops = new HashSet<>();
+  private final Set<String> seenBicycleRentalStations = new HashSet<>();
+  private final boolean includeStops;
+  private final boolean includePatternAtStops;
+  private final boolean includeBikeShares;
+  private final int maxResults;
+
+  /**
+   *
+   * @param routingService A RoutingService used in finding information about the various places.
+   * @param filterByModes A list of TransitModes for which to find Stops and PatternAtStops. Use null to disable the filtering.
+   * @param filterByPlaceTypes A list of PlaceTypes to search for. Use null to disable the filtering, and search for all types.
+   * @param filterByStops A list of Stop ids for which to find Stops and PatternAtStops. Use null to disable the filtering.
+   * @param filterByRoutes A list of Route ids used for filtering Stops. Only the stops which are served by the route are returned. Use null to disable the filtering.
+   * @param filterByBikeRentalStations A list of BikeRentalStation ids to use in filtering.  Use null to disable the filtering.
+   * @param maxResults Maximum number of results to return.
+   */
+  public PlaceFinderTraverseVisitor(
+      RoutingService routingService, List<TransitMode> filterByModes,
+      List<PlaceType> filterByPlaceTypes, List<FeedScopedId> filterByStops,
+      List<FeedScopedId> filterByRoutes, List<String> filterByBikeRentalStations, int maxResults
+  ) {
+    this.routingService = routingService;
+    this.filterByModes = toSet(filterByModes);
+    this.filterByStops = toSet(filterByStops);
+    this.filterByRoutes = toSet(filterByRoutes);
+    this.filterByBikeRentalStation = toSet(filterByBikeRentalStations);
+
+    includeStops = filterByPlaceTypes == null || filterByPlaceTypes.contains(PlaceType.STOP);
+    includePatternAtStops = filterByPlaceTypes == null
+        || filterByPlaceTypes.contains(PlaceType.PATTERN_AT_STOP);
+    includeBikeShares = filterByPlaceTypes == null
+        || filterByPlaceTypes.contains(PlaceType.BICYCLE_RENT);
+    this.maxResults = maxResults;
+  }
+
+  private static <T> Set<T> toSet(List<T> list) {
+    if (list == null) { return null; }
+    return new HashSet<T>(list);
```

Comment:
Tip: Set.copyOf(list) is preferable since there is no modifications to the set afterwards.

Reply:
Changed